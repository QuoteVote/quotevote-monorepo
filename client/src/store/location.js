/**
 * Location Redux Store
 * Manages user's current location and permission state
 */

import { createSlice } from '@reduxjs/toolkit';
import { getCurrentPosition, checkPermission, GeoLocationError } from '../utils/geolocationService';

const initialState = {
  currentLocation: null, // { lat, lng }
  permissionState: 'prompt', // 'prompt' | 'granted' | 'denied' | 'unsupported'
  isLoading: false,
  error: null, // { type, message }
  lastUpdated: null,
};

const locationSlice = createSlice({
  name: 'location',
  initialState,
  reducers: {
    setLocation: (state, action) => {
      state.currentLocation = action.payload;
      state.lastUpdated = Date.now();
      state.error = null;
    },
    setPermissionState: (state, action) => {
      state.permissionState = action.payload;
    },
    setLoading: (state, action) => {
      state.isLoading = action.payload;
    },
    setError: (state, action) => {
      state.error = action.payload;
      state.isLoading = false;
    },
    clearError: (state) => {
      state.error = null;
    },
    clearLocation: (state) => {
      state.currentLocation = null;
      state.lastUpdated = null;
      state.error = null;
    },
    reset: () => initialState,
  },
});

export const {
  setLocation,
  setPermissionState,
  setLoading,
  setError,
  clearError,
  clearLocation,
  reset,
} = locationSlice.actions;

// Thunks

/**
 * Fetch user's current location
 * @param {object} options - Options passed to getCurrentPosition
 */
export const fetchLocation = (options = {}) => async (dispatch) => {
  dispatch(setLoading(true));
  dispatch(clearError());

  try {
    const location = await getCurrentPosition(options);
    dispatch(setLocation(location));
    dispatch(setPermissionState('granted'));
    return location;
  } catch (error) {
    dispatch(setError({
      type: error.type || GeoLocationError.UNKNOWN,
      message: error.message || 'Failed to get location',
    }));
    
    if (error.type === GeoLocationError.PERMISSION_DENIED) {
      dispatch(setPermissionState('denied'));
    }
    
    throw error;
  } finally {
    dispatch(setLoading(false));
  }
};

/**
 * Check and update permission state
 */
export const updatePermissionState = () => async (dispatch) => {
  try {
    const state = await checkPermission();
    dispatch(setPermissionState(state));
    return state;
  } catch (error) {
    console.error('Failed to check permission:', error);
    return 'prompt';
  }
};

// Selectors

export const selectLocation = (state) => state.location.currentLocation;
export const selectPermissionState = (state) => state.location.permissionState;
export const selectIsLoading = (state) => state.location.isLoading;
export const selectError = (state) => state.location.error;
export const selectLastUpdated = (state) => state.location.lastUpdated;

export const selectHasLocation = (state) => state.location.currentLocation !== null;
export const selectIsPermissionGranted = (state) => state.location.permissionState === 'granted';
export const selectIsPermissionDenied = (state) => state.location.permissionState === 'denied';

export default locationSlice.reducer;
