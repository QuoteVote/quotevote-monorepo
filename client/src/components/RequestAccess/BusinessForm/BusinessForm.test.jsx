import React from 'react'
import { render, act } from '@testing-library/react'
import { vi } from 'vitest'
import { Provider } from 'react-redux'
import BusinessForm from './BusinessForm'
import store from '../../../store/store'

// PaymentMethod triggers async updates and requires many props. Mock it
// to avoid act() warnings and PropTypes noise for this shallow unit test.
vi.mock('../PaymentMethod/PaymentMethod', () => {
  const React = require('react')
  return {
    __esModule: true,
    default: (props) => React.createElement('div', { 'data-testid': 'mock-payment-method', ...props }),
  }
})

const props = {
  requestInviteSuccessful: false,
  handleSubmit: vi.fn(),
  register: vi.fn(),
  errors: {},
  isContinued: false,
  onContinue: vi.fn(),
  setCardDetails: vi.fn(),
  cardDetails: { cardNumber: '', expiry: '', cvv: '', cost: 0 },
  onSubmit: vi.fn(),
  errorMessage: null,
  loading: false,
}

const BusinessFormWrapper = () => (
  <Provider store={store}>
    <BusinessForm {...props} />
  </Provider>
)

describe('BusinessForm test -', () => {
  it('renders correctly', async () => {
    let container
    await act(async () => {
      const res = render(<BusinessFormWrapper />)
      container = res.container
    })
    // PaymentMethod is mocked for this unit test; avoid asserting fragile
    // markup snapshots and opt for a stable existence check.
    expect(container.firstChild).toBeTruthy()
  })
})
