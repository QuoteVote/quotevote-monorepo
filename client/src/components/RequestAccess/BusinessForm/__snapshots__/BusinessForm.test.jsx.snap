// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BusinessForm test - > renders correctly 1`] = `
<div
  class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root"
  justify="center"
  style="margin-right: 24px;"
>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter makeStyles-header-2 css-d4fp74-MuiTypography-root"
    >
      Get access to your
       
      <span
        class="makeStyles-header-2"
        style="color: rgb(82, 178, 116);"
      >
        Business Plan!
      </span>
    </p>
  </div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-alignCenter makeStyles-subHeader-3 css-d4fp74-MuiTypography-root"
    >
      You are one step away from
       
      <b>
        unlimited access
      </b>
       
      to Quote Vote
    </p>
  </div>
  <div
    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
    style="margin-top: 2%;"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row css-1rzj21q-MuiGrid-root"
        justify="center"
      >
        <img
          alt="Business Illustration"
          height="500"
          src="/assets/Illustration.png"
          style="width: 400px; height: 265px; object-fit: contain;"
        />
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-yu7yff-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-6fm6rz-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
            >
              <div
                class="MuiCardHeader-avatar css-1r9wl67-MuiCardHeader-avatar"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 makeStyles-stepNumber-4 css-1edfpdg-MuiTypography-root"
                >
                  1
                </p>
              </div>
              <div
                class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
                  style="font: Roboto; line-height: 1.56;"
                >
                  Your Personal Info
                </p>
              </div>
            </div>
            <form>
              <div
                class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-2 css-zjhuaj-MuiGrid-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4n9bg-MuiFormControl-root-MuiTextField-root"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1htra48-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for="fullName"
                        id="fullName-label"
                      >
                        Full Name
                        <span
                          aria-hidden="true"
                          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                        >
                           
                          *
                        </span>
                      </label>
                      <div
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1bz4o94-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-invalid=""
                          class="MuiInputBase-input MuiOutlinedInput-input css-165z7mv-MuiInputBase-input-MuiOutlinedInput-input"
                          id="fullName"
                          name="fullName"
                          required=""
                          type="text"
                          value=""
                        />
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1ddm75b-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-1cs9q92-MuiNotchedOutlined-root"
                          >
                            <span>
                              Full Name
                               
                              *
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                      
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4n9bg-MuiFormControl-root-MuiTextField-root"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1htra48-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for="companyName"
                        id="companyName-label"
                      >
                        Company Name
                        <span
                          aria-hidden="true"
                          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                        >
                           
                          *
                        </span>
                      </label>
                      <div
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1bz4o94-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiOutlinedInput-input css-165z7mv-MuiInputBase-input-MuiOutlinedInput-input"
                          id="companyName"
                          name="companyName"
                          required=""
                          type="text"
                          value=""
                        />
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1ddm75b-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-1cs9q92-MuiNotchedOutlined-root"
                          >
                            <span>
                              Company Name
                               
                              *
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-y4n9bg-MuiFormControl-root-MuiTextField-root"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-1htra48-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="false"
                        for="email"
                        id="email-label"
                      >
                        Email
                        <span
                          aria-hidden="true"
                          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1ljffdk-MuiFormLabel-asterisk"
                        >
                           
                          *
                        </span>
                      </label>
                      <div
                        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1bz4o94-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiOutlinedInput-input css-165z7mv-MuiInputBase-input-MuiOutlinedInput-input"
                          id="email"
                          name="email"
                          required=""
                          type="text"
                          value=""
                        />
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1ddm75b-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-1cs9q92-MuiNotchedOutlined-root"
                          >
                            <span>
                              Email
                               
                              *
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
                  >
                    <button
                      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary makeStyles-greenBtn-1 css-g1oxnb-MuiButtonBase-root-MuiButton-root"
                      tabindex="0"
                      type="submit"
                    >
                      Continue
                    </button>
                  </div>
                </div>
              </div>
            </form>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-direction-xs-row css-17fpwt7-MuiGrid-root"
        >
          <form>
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-6fm6rz-MuiPaper-root-MuiCard-root"
            >
              <div
                class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
              >
                <div
                  class="MuiCardHeader-avatar css-1r9wl67-MuiCardHeader-avatar"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 makeStyles-stepNumber-24 css-1edfpdg-MuiTypography-root"
                  >
                    2
                  </p>
                </div>
                <div
                  class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 makeStyles-stepName-25 css-1edfpdg-MuiTypography-root"
                  >
                    Would you like to support this app?
                  </p>
                </div>
              </div>
              <input
                name="cardNumber"
                type="hidden"
              />
              <input
                name="expiry"
                type="hidden"
              />
              <input
                name="cvv"
                type="hidden"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</div>
`;
